### PR TITLE
Port controller to Golang

### PR DESCRIPTION
**Pull request by caseydavenport**
## Description

This PR introduces a golang implementation of the policy/pod/namespace controller.

Fixes: https://github.com/projectcalico/k8s-policy/issues/119
Fixes: https://github.com/projectcalico/k8s-policy/issues/98

## Todos

- [x] Unit tests
- [ ] Release note

## Release Note

```release-note
`calico/kube-policy-controller` has been ported to golang
```

### CODE REVIEW

File: glide.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package: github.com/projectcalico/k8s-policy
+description: Calico Policy Controller
+homepage: https://projectcalico.org
+license: Apache-2.0
+import:
+- package: github.com/sirupsen/logrus
+  version: ^0.10.0
+- package: github.com/kelseyhightower/envconfig
+  version: ~1.3.0
+- package: github.com/kelseyhightower/envconfig
```
Comment (gunjan5) [inline glide.yaml]: why do we have this one twice?

File: Makefile [hunk 2]
```diff
@@ -1,27 +1,102 @@
...
+	-mkdir -p .go-pkg-cache
+	docker run --rm --privileged --net=host \
+		-e LOCAL_USER_ID=$(LOCAL_USER_ID) \
+		-v $(CURDIR)/.go-pkg-cache:/go/pkg/:rw \
+		-v $(CURDIR):/go/src/$(PACKAGE_NAME):rw \
+		$(CALICO_BUILD) sh -c 'cd /go/src/$(PACKAGE_NAME) && make WHAT=$(WHAT) SKIP=$(SKIP) ut'
 
-# Run system tests.
+GET_CONTAINER_IP := docker inspect --format='{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}'
+K8S_VERSION=1.7.4
```
Comment (gunjan5) [inline Makefile]: BTW, k8s 1.7.5 is out. Wish hyperkube had a `latest` or a `dev` tag so we can keep testing the latest k8s versions

File: pkg/cache/cache.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+	// Indicate that the cache is running, and so updates
+	// can be queued.
+	c.mut.Lock()
+	c.running = true
+	c.mut.Unlock()
+}
+
+func (c *calicoCache) isRunning() bool {
+	c.mut.Lock()
```
Comment (gunjan5) [inline pkg/cache/cache.go]: do we really need a mutex lock for reading?

File: pkg/cache/cache_test.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache_test
+
+import (
+	"fmt"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/projectcalico/k8s-policy/pkg/cache"
```
Comment (gunjan5) [inline pkg/cache/cache_test.go]: 🌮

File: pkg/cache/cache.go (hunk 3 above)
Comment (caseydavenport) [reply pkg/cache/cache.go]: Probably not tbh, but I was rushing and was being safe :)

File: pkg/cache/cache_test.go (hunk 4 above)
Comment (caseydavenport) [reply pkg/cache/cache_test.go]: ek

File: Makefile (hunk 2 above)
Comment (caseydavenport) [reply Makefile]: Yeah.  For these sorts of tests the patch releases don't matter much, but the minor releases are more important.

**Pull request merged**